&copy; Mirko Piani - *PhD student in Precision Orchard Management (POM), university of Bologna (IT)*
___

If you are having problems with the notebook, contact me at mirko.piani2@unibo.it

# Python Operators

Operators are special symbols in python that carry out arthimetic and logical operations.

In [2]:
# example
2 + 5

7

in this case, the operator symbol `+` is allowing to sum the two integer numbers

## Operator Types

There are different types of operators:
1. Arithmetic operators: 
   - *addition* `+`,
   - *subtraction* `-`,
   - *multiplication* `*`,
   - *floating-point division* `/`,
   - *floor-division* `//`,
   - *exponentiation* `**`.
  
2. Comparison operators
   - *is equal* `==`,
   - *is different* `!=`,
   - *is lower than* `<`,
   - *is greater than* `>`,
   - *is lower or equal to* `<=`,
   - *is greater or equal to* `>=`.
  
3. Logical operators
   - *AND*,```and``` - gives `True` if both operands are `True`
   - *OR*, ```or``` - gives `True` if at least one operand is `True`
   - *NOT*,```and not``` - gives `True` if operand is `False`

4. Bitwise operators
   - *Bitwise AND*, `&`
   - *Bitwise OR*, `|`
   - *Bitwise NOT*, `~`
   - *Bitwise XOR*, `^`
   - *Bitwise Right Shift*, `>>`
   - *Bitwise Left Shift*, `<<`

5. Assignment operators
   - *assignment operator*, `=`
   - *addition assignment operator*, `+=`
   - *subtraction assignment operator*, `-=`
   - *multiplication assignment operator*, `*=`
   - *division assignment operator*, `/=`
   - *floor-division assignment operator*, `//=`
   - *exponentiation assignment operator*, `**=`

6. Special operators

   - `is` - gives `True` if the operands are identical
   - `is not` - gives `True` if the operands are not identical
   - `in` - gives `True` if value / variable is found in sequence
   - `not in` - gives `True` if value / variable is not found in the sequence


## Arithmetic operators

In [4]:
x = 15
y = 6

# Addition Operator
print(f'x + y = { x + y}')

# Subtraction Operator
print(f'x - y = { x - y}')

# Multiplication Operator
print(f'x * y = { x * y} ')

# Division Operator
print(f'x / y = { x / y} ') 
print(f'x // y = { x // y}')

# Exponential Operator
print(f'x ** y = { x ** y}')


x + y = 21
x - y = 9
x * y = 90 
x / y = 2.5 
x // y = 2
x ** y = 11390625


## Comparison operators

In [5]:
x = 1
y = 6

# Equal To Operator
print(f'x == y = { x == y}')

# Not Equal To Operator
print(f'x != y = { x != y}')

# Greater Than Operator
print(f'x > y = { x > y}')

# Greater Than or Equal To
print(f'x >= y = { x >= y}')

# Lesser Than Operator
print(f'x < y = { x < y} ')

# Lesser Than or Equal To
print(f'x <= y = { x <= y} ') 


x == y = False
x != y = True
x > y = False
x >= y = False
x < y = True 
x <= y = True 


## Logical operators

In [9]:
var1 = True
var2 = False

# and
print(f'var1 and var2 = { var1 and var2}')

# or
print(f'var1 or var2 = { var1 or var2}')

# not
print(f'var1 and not var2 = { var1 and not var2}')

var1 and var2 = False
var1 or var2 = True
var1 and not var2 = True


## Bitwise operators

In [11]:
x = 10
y = 5

# Bitwise AND
print(f'x & y = { x & y}')

# Bitwise OR
print(f'x | y = { x | y}')

# Bitwise NOT
print(f'x != y = { x != y}')

# Bitwise XOR
print(f'x ^ y = { x ^ y}')

# Right Shift
print(f'x >> y = { x >> y} ')

# Left Shift
print(f'x << y = { x << y} ') 


x & y = 0
x | y = 15
x != y = True
x ^ y = 15
x >> y = 0 
x << y = 320 


What is the logic behind that? how is it possible to get `15` or `320` from `10` and `5`?

*8bit Binary numbers*

| $ 2^7 $ | $ 2^6 $ | $ 2^5 $ | $ 2^4 $ | $ 2^3 $ | $ 2^2 $ | $ 2^1 $ | $ 2^0 $ | 

Binary representation:
- `x` (10) in binary: `1010`
- `y` (5) in binary: `0101`

Thus,

| value | $ 2^7 $ | $ 2^6 $ | $ 2^5 $ | $ 2^4 $ | $ 2^3 $ | $ 2^2 $ | $ 2^1 $ | $ 2^0 $ |
| ---|---|---|---|---|---|---|---|---|
| 10 | 0 | 0 | 0 | 0 | **1** | **0** | **1** | **0** |
| 5 | 0 | 0 | 0 | 0 | **0** | **1** | **0** | **1** |


**Bitwise OR**

```python
print(f'x | y = { x | y}')

```

the code above checks at bit location if a variable has a `True` (i.e. is equal to 1), if it is,it returns 1 at that bit location. Thus, it returns:

| value | $ 2^7 $ | $ 2^6 $ | $ 2^5 $ | $ 2^4 $ | $ 2^3 $ | $ 2^2 $ | $ 2^1 $ | $ 2^0 $ |
| ---|---|---|---|---|---|---|---|---|
| 10 | 0 | 0 | 0 | 0 | **1** | **0** | **1** | **0** |
| 5 | 0 | 0 | 0 | 0 | **0** | **1** | **0** | **1** |
| **output** | 0 | 0 | 0 | 0 | **1** | **1** | **1** | **1** |


The binary `1111` is $ 8 (2^3) + 4 (2^2) + 2 (2^1) + 1 (2^0) = 15$


**Left Shift** - shift the first number `x` to the left by `y` times `0`

```python
print(f'x << y = { x << y}')

```

the code above checks at bit location if a variable has a `True` (i.e. is equal to 1), if it is,it returns 1 at that bit location. Thus, it returns:

**Left Shift** 
```yaml
    Before left shift:  1010
    After left shift: 101000000

```

Thus: 
```scss

(1 * 2^8) + (0 * 2^7) + (1 * 2^6) + (0 * 2^5) + (0 * 2^4) + (0 * 2^3) + (0 * 2^2) + (0 * 2^1) + (0 * 2^0)
= (1 * 256) + (0 * 128) + (1 * 64) + (0 * 32) + (0 * 16) + (0 * 8) + (0 * 4) + (0 * 2) + (0 * 1)
= 256 + 0 + 64 + 0 + 0 + 0 + 0 + 0 + 0
= 320


```

## Assignment operators

In [17]:
a = 5
print(f'starting value: {a}')

a += 5  # take the value of a, add 5 to it, and reassign the output to variable a. Thus a = a + 5
print(a)

a -= 5  # a = a - 5 
print(a)

a *= 5  # a = a * 5
print(a)

a /= 5  # a = a / 5
print(a)

a //= 2  # a = a // 2
print(a)

a %= 1  # a = a % 1
print(a)

a = 10
a **= 2  # a = a ** 2
print(a)

starting value: 5
10
5
25
5.0
2.0
0.0
100


## Special operators

In [18]:
x1 = 2  # int data type
y1 = 2  # int
x2 = 'Hello'  # string
y2 = "Hello"  # string
x3 = [1,2,3]  # list
y3 = (1,2,3)  # tuple

print('x1 is y1 = ', x1 is y1)
print('x1 is y2 = ', x1 is y2)
print('x3 is not y3 = ', x3 is not y3)

x1 is y1 =  True
x1 is y2 =  False
x3 is not y3 =  True


In [15]:
x = 'Hello World'
y = {1:'a', 2:'b'}

print("'H' in x ", 'H' in x)
print('hello not in x ','hello' not in x)
print('1 in y = ', 1 in y)
print('a in y = ', a in y)

'H' in x  True
hello not in x  True
1 in y =  True
a in y =  False
